### ISSUE TITLE
[FR] Colorbar properties

### ISSUE THREAD

**Issue by rafaqz** [JuliaPlots/Plots.jl#3174]
It would be useful to be able to style additional attributes of colorbars, as discussed recently on slack in plotting

~- [ ] tick size~
~- [ ] tick color~
- [x] ticks
- [x] tick scale
- [x] tick font
- [x] tick font size
- [x] tick font color
- [ ] border color
- [ ] border thickness
- [ ] bar width
- [ ] bar height
- [ ] other things?

My use-case is cleaner map-plotting recipes in GeoData.jl. Currently, if you stray too far from the defaults the colorbar starts to look out of place.

This could start with just GR?

related, but just for ticks: https://github.com/JuliaPlots/Plots.jl/issues/2308

**Comment by mkborregaard** [comment]
It's tricky because Plots development aims to keep keywords few(-ish), in order to keep maintainability. In general all new keywords should preferably be applied across all backends to keep interchangeability of backends. One possibility could be to adapt `extra_kwargs` to be able to pass colorbar kwargs, but this wouldn't work in recipes. I guess this requires a bit of discussion.

@rafaqz do you happen to know how many of these attributes are supported natively by e.g. GR?

**Comment by hervasa2** [comment]
I think the colorbar should have the same attributes as an axis. For example just like you have xticks and xscale, the colorbar should be able to support cticks and cscale. I came across this limitation while attempting cscale = :log10 (also tried zscale = :log10) on a heatmap.

**Comment by rafaqz** [comment]
I can see how keeping the keyword scope down would be important! especially given the breadth of use cases plots has to deal with. 

If we can use the same set of keywords (maybe with `colorbar_` prefixed?) it would at least add less overhead in terms of new names for things.

I'm not sure what is possible with GR, the `colorbar` method doesn't have many arguments, but there is lot of global state that is incorporated somehow. It does seem possible to do with matplotlib, plotly, and pgfplots.

**Comment by mkborregaard** [comment]
So having a new keyword type `colorbar` in addition to `Plot`, `Subplot`, `Axis` and `Series`, and reusing `Axis` keywords while prepending `colorbar_`? That does sound like a neat design. Thoughts @daschw ?

**Comment by rafaqz** [comment]
Also I didn't read @hervas2 comment properly, `cticks` etc is nice too for where x/y is usually used.

**Comment by daschw** [comment]
> So having a new keyword type colorbar in addition to Plot, Subplot, Axis and Series, and reusing Axis keywords while prepending colorbar_? That does sound like a neat design. Thoughts @daschw ?

I think that sounds like a good approach.

**Comment by hervasa2** [comment]
Yes colorbar_[keyword] with c[keyword] as an alias would be nice and intuitive. I think a minimal feature in GR (if not all axis attributes can be applied to the colorbar) would be colorbar_ticks, that way the user could manipulate the data behind the scenes and set manual ticks to achieve the desired result.

**Comment by isentropic** [comment]
I mean there is `clims` already, ticks would be more difficult for sure

**Comment by fhagemann** [comment]
I would also very much welcome the ability of manipulating colorbar properties. Thanks for bringing this up!

**Comment by tbeason** [comment]
I'm also looking for this functionality. Asked for exactly this in Slack and was directed here

**Comment by bcsj** [comment]
I think it would be nice to be able to access the frame/border/axes of the colorbar as well, e.g. `colorbar_frame/cframe=:box`.

**Comment by Gyoshi** [comment]
related: #2345

**Comment by isentropic** [comment]
If anyone is willing to help out with the development of this, please help out with the testing in the PR

**Comment by isentropic** [comment]
@rafaqz and others, most of the requested changes have been merged, please let me know your comments. Currently only PyPlot backend supports the new colorbar args:
```
    :colorbar_fontfamily        => :match,
    :colorbar_ticks              => :auto,
    :colorbar_tickfontfamily    => :match,
    :colorbar_tickfontsize      => 8,
    :colorbar_tickfontcolor     => :match,
    :colorbar_scale             => :identity,
```

**Comment by isentropic** [comment]
currently no c_*** args are made yet, but let's first make sure this works

**Comment by rafaqz** [comment]
Thanks, I'll try to give some feedback in the next few days

**Comment by isentropic** [comment]
Naturally you'd need to use #master branch before we get a relase

**Comment by rafaqz** [comment]
These work perfectly for me. Although I'm not sure what the allowed values for `colorbar_scale` are?

It would be good to add a few controls for the colorbar border too, like color/width etc. Seems it may be easy to copy your PR now to add some related args :)

**Comment by isentropic** [comment]
colorbar_scale takes :log10, :log2 and :identity (same as axis). Regarding the border and the color, it is strange that you want to change it even. Right now pyplot has 5% border width of the main plot, and I have no idea what do you wish we could've controlled

**Comment by rafaqz** [comment]
Oh right yep scale works well too.

With color/width - there are situations where you don't want a border on the colorbar, or you want it to be less prominent. For example, if you use e.g. `:border=:none` the plot has no border, but the colorbar still has one, and it doesn't look great. You may also want to use e.g. a ligher grey border instead of black, or use a thinner border, to push the colorbar back and make the plot stand out more. That being said, the PyPlot colorbar does look alright as-is most of the time, GR less so.

But I can add this later.

**Comment by hervasa2** [comment]
Thank you for working on this! `colorbar_scale` is working nicely; however now the problem is the color gradient itself. When the `colorbar_scale` is switched to `:log10` the color gradient follows it likewise. Therefore, once `colorbar_scale = :log10` is set only the colorbar changes, and not the actual colors on the heatmap.  When I tried to manually fix this by using something like `cgrad(:inferno, scale =:exp)` I came across the bug reported on #3420. pyplot no longer respects a color gradient's scale

**Comment by isentropic** [comment]
Right, I've been aware of this for a while.
The attempt to fix was here #3432. I need some additional testers

**Comment by isentropic** [comment]
Right now this is a pyplot issue right? I just want to make sure that simplest behavior is still functional. Please let me know if

1. Normal linear scale colorbars work as intended with pyplot
2. Markers are not properly colored with the log scale colorbar.

**Comment by hervasa2** [comment]
Thanks for your quick response! Yes this is a pyplot issue only. 

1. Normal linear scale colorbars work as intended with pyplot. Base functionality looks fine.
2. Markers are not properly colored with the log scale colorbar. This applies both to the cases where one uses `colorbar_scale = :log` or when `color = cgrad(:inferno, scale =:log)`

In the meantime I will test  #3432

**Comment by hervasa2** [comment]
I think the first order of business would be to get `color = cgrad(:inferno, scale =:log)` working again since this is something that was broken. 
Then I like the axis formatter approach suggested by @isentropic in #3432:

"Can't Plots just provide axis formatters?
For example:
I am plotting a 10^2, 10^3, 10^4 (three points) in log scale,
Then, Plots would just log it to get -> (2, 3, 4)
Then, just decorate the axis labels as (10^2, 10^3, 10^4) (instead of natural 2,3,4)"

**Comment by isentropic** [comment]
I'm not sure how to handle colorbars with log scale right now. I don't think we should focus on (cgrad=:log) just yet.
In the pr Cscale=log produces normal looking colorbar (color transition in the gradient looks uniform). However the markers are wrong color. 



The problem is that I'm not so sure where Plots assigns marker colors to the points . If someone has an idea do let me know.

**Comment by BeastyBlacksmith** [comment]
Marker colors are stored in the `marker_z` seriesattribute, but maybe I got the question wrong

**Comment by isentropic** [comment]
Yes, I'm aware of that, but where does it get assigned? Somewhere in the depths of the pipeline but  I could not find where

**Comment by BeastyBlacksmith** [comment]
Aren't those given from the callsite? Or do you have examples where there is automatic color assignment?

**Comment by isentropic** [comment]
In case of the colorbar, I think the colors need to be assigned somehow? Perhaps I'm missing something here but I thought marker color gets overriten when a colobar is present

**Comment by BeastyBlacksmith** [comment]
Its the other way around. You get a colorbar of you got marker_z values. How should Plots now what values to put there?

**Comment by isentropic** [comment]
@hervasa2 @rafaqz checkout the PR 
https://github.com/JuliaPlots/Plots.jl/pull/3620

**Comment by isentropic** [comment]
need testing and feedback from you

**Comment by zhiyuanzhai** [comment]
Other backend are still to support the `colorer_scale` feature. For example, with PGFPlotsX:
<img width="110" alt="image" src="https://user-images.githubusercontent.com/37478391/162885530-2ff2fc1f-394e-4b1e-91c9-42f015bd575e.png">

**Comment by CantelopePeel** [comment]
any plan to propagate the log colorbar support to other backends?

**Comment by t-bltg** [comment]
It should work for the `pythonplot` backend after https://github.com/JuliaPlots/Plots.jl/pull/4542.

**Comment by cmdenis** [comment]
Has there been any updates in adding these functionalities to the `GR` backend?

**Comment by alex180500** [comment]
> I think it would be nice to be able to access the frame/border/axes of the colorbar as well, e.g. `colorbar_frame/cframe=:box`.

are there news on how to implement this?

**Comment by BeastyBlacksmith** [comment]
Note: Goal of this issue is to have full colorbar support for all of the backends. So in order to claim the bounty, the colorbar control should be implemented at least for GR, plotly/plotlyjs and PGFPlotsX.
